test/framework/src/main/java/org/elasticsearch/transport/nio/TcpReadHandler.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio;
+
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.transport.nio.channel.NioSocketChannel;
+
+import java.io.IOException;
+
+public class TcpReadHandler {
+
+    private final NioTransport transport;
+
+    public TcpReadHandler(NioTransport transport) {
+        this.transport = transport;
+    }
+
+    public void handleMessage(BytesReference reference, NioSocketChannel channel, String profileName,
+                              int messageBytesLength) {
+        try {
+            transport.messageReceived(reference, channel, profileName, channel.getRemoteAddress(), messageBytesLength);
+        } catch (IOException e) {
+            handleException(channel, e);
+        }
+    }
+
+    public void handleException(NioSocketChannel channel, Exception e) {
+        transport.exceptionCaught(channel, e);
+    }
+
+    public void handlePing(NioSocketChannel channel, String profileName) {
+
```

Comment:
More of like a "possibly todo". Currently it seems like we only do anything with pings on the "send" side. I had added a handler in case we ever wants to trace or keep metrics about or anything from the receive side. I'll think about it a bit but I'll probably remove it.